warehouse/forklift/legacy.py

```diff
@@ -394,6 +394,40 @@ def _validate_classifiers(form, field):
             )
 
 
+def _validate_dynamic(_form, field):
+    declared_dynamic_fields = {str.title(k) for k in field.data or []}
+    disallowed_dynamic_fields = {"Name", "Version", "Metadata-Version"}
+    if invalid := (declared_dynamic_fields & disallowed_dynamic_fields):
+        raise wtforms.validators.ValidationError(
+            f"The following metadata field(s) cannot be marked as dynamic: {invalid!r}",
+        )
+    # "Supported-Platform" is missing in the metadata form since does not have any
+    # constraints
+    allowed_dynamic_fields = (
+        {field.description for field in MetadataForm() if field.description}
+        | {"Supported-Platform"}
```

Comment:
I understand why this is here but this seems a bit weird, perhaps we should add the field to the form instead, even if we don't do anything with it.

It also seems like a bit of a hack that we're relying on the description for these values, should we be getting this from `DynamicFieldsEnum` instead?

Reply:
> I understand why this is here but this seems a bit weird, perhaps we should add the field to the form instead, even if we don't do anything with it.

Done

> It also seems like a bit of a hack that we're relying on the description for these values, should we be getting this from DynamicFieldsEnum instead?

I think we had that and changed it due to https://github.com/pypi/warehouse/pull/13606#discussion_r1261291228, or am i misreading your comment?

Reply:
What I meant there is that I don't think we don't need to check if `Dynamic` contains fields that aren't valid metadata fields (e.g. `Dynamic: ThisDoesntExist`), as those will fail when they're provided as metadata fields (e.g. `ThisDoesntExist: Foo`) *but* that assumes that the validator will always fail when `Dynamic` contains any value that isn't also present as a valid metadata field (e.g., `Dynamic: Author`, but `Author` isn't provided).

If the latter is true, we don't need the Enum or to check if `Dynamic` contains a valid metadata field name, we only need to:

1. Check that it doesn't contain a disallowed field name (`Dynamic: Metadata-Version` should fail);
2. Check that it only contains field names that are also provided as metadata (`Dynamic: Author` should fail if `Author` is not provided, `Dynamic: ThisDoesntExist` should fail because the form prevents `ThisDoesntExist` from being provided).

Reply:
OK, that makes sense too! Given that I think we do need the Enum but we can simplify this, I'll provide a suggestion in a moment.